Rewrite in functional style

spawnia opened the pull request:
During todays attempt to implement https://github.com/DiederikvandenB/apollo-link-sentry/pull/265, i found it increasingly difficult to
- work within the class-based implementation
- deal with the classes that are passed to the callbacks defined within the options

In my experience, plain objects and (mostly) pure functions lead to simpler code that is easier to work with, less stateful and less error prone. What do you think about this approach?

This PR is just a draft as of now. If you don't like it, that is totally fine, i can also maintain it in my fork. Perhaps, we can find common ground and continue working on this together. Next steps for me would be to add spans and some performance tracing.

In any case thank you for your previous work on this, i found the general structure and options to really make sense.

DiederikvandenB commented:
I wanted to look at this today but I didn’t find the time for it. I am definitely open for refactoring and I’ll have a closer look after the weekend.

spawnia commented on README.md:
```diff
@@ -39,79 +39,96 @@ const client = new ApolloClient({
+   *
+   * If undefined, all operations will be included.
+   */
+  shouldHandleOperation: undefined | ((operation: Operation) => boolean);
```
This definition style:
- shows the possible options
- indicates the default
- allows for a single point of truth that can also be navigated to from inside the IDE

spawnia replied:
I found the name `filter` to require explanation as to what it does.

spawnia commented on README.md:
```diff
@@ -39,79 +39,96 @@ const client = new ApolloClient({
+   * }
    */
-  setFingerprint: true,
+  attachBreadcrumbs: AttachBreadcrumbsOptions | false;
```
Perhaps this should also be singular?

DiederikvandenB replied:
I guess an error contains multiple breadcrumbs, but whatever you want I'm also fine with.

spawnia replied:
Yeah, i think this reads more natural. Thanks for the feedback.

spawnia commented on README.md:
```diff
@@ -39,79 +39,96 @@ const client = new ApolloClient({
-   * data to be included. Very useful in combination with options like includeVariables and includeContextKeys.
-   */
-  beforeBreadcrumb: (breadcrumb) => breadcrumb,
+    transform:
```
Moved in here from a top level config, only makes sense when breadcrumbs are generally enabled.

spawnia commented on README.md:
```diff
@@ -39,79 +39,96 @@ const client = new ApolloClient({
-     * option is toggled, the breadcrumb will only show the operation name and it's type.
+     * Include the full query string?
      */
-    enable: true,
```
Replaced by setting the top level `attachBreadcrumbs: false`.

spawnia commented on src/apolloSentryLink.ts:
```diff
@@ -0,0 +1,89 @@
+  setTransaction,
+} from './sentry';
+
+export function makeSentryLink(options: Options = {}): ApolloLink {
```
Still unsure if this would be nicer as a class.

spawnia replied:
Reverted to the class based API to avoid unnecessary breakage.

spawnia commented on src/apolloSentryLink.ts:
```diff
@@ -0,0 +1,89 @@
+          fullOptions.attachBreadcrumbs.includeResult
+        ) {
+          // We must have a breadcrumb if attachBreadcrumbs was set
+          (breadcrumb as Breadcrumb).data.result = result;
```
Perhaps this can be done in a more elegant way. Orginally, I had a check for `&& breadcrumb` above, but it is actually unnecessary. This version has no cost at runtime.

DiederikvandenB replied:
Agree it's not the most beautiful way, but it's also not a huge issue for me.

spawnia commented on README.md:
```diff
@@ -39,79 +39,96 @@ const client = new ApolloClient({
+   *
+   * The following information will be included by default:
+   * {
+   *   type: 'http',
```
I found it unnecessary to switch the type to `error`, since that is what `level` is for. Perhaps there are situtations where GraphQL does not go over HTTP, should we account for those?

DiederikvandenB replied:
It's been too long to remember why I set the type to error. I would assume it had something to do with the way the error is rendered in the Sentry dashboard.

I'm not currently working on a project which uses Sentry, so I can't test that out right now. If everything shows like it's supposed to, I'm fine with this change too.

spawnia replied:
![image](https://user-images.githubusercontent.com/12158000/107258704-692beb00-6a3c-11eb-881f-5bc939e323c5.png)

I think this looks really nice.

DiederikvandenB replied:
Agreed, looks good.

spawnia commented on README.md:
```diff
@@ -39,79 +39,96 @@ const client = new ApolloClient({
+   * The following information will be included by default:
+   * {
+   *   type: 'http',
+   *   category: `graphql.${operationType}`,
```
This fits with the default breadcrumbs of the browser integration, e.g. `ui.click`.

spawnia commented on src/breadcrumb.ts:
```diff
@@ -0,0 +1,69 @@
+export interface BreadcrumbData {
+  query?: string;
+  variables?: Record<string, unknown>;
+  result?: FetchResult;
```
Perhaps `fetchResult` would be nicer?

DiederikvandenB replied:
I agree

spawnia replied:
Changed.

spawnia commented on src/operation.ts:
```diff
@@ -0,0 +1,27 @@
+export function makeOperation(apolloOperation: ApolloOperation): Operation {
+  const { query, operationName } = apolloOperation;
+
+  // We know this can not be null because Apollo validates the document/operationName before
```
I am actually not entirely sure if they do, will try it out soon.

spawnia replied:
I tried it out, they actually don't even allow multiple operations within a doc.

spawnia commented:
Not quite ready to merge, but complete enough to be worth a review.

DiederikvandenB commented:
Thanks for taking the time to write this PR.

I mostly decided to go the class way because Sentry's class based too, but I like your approach as well. Let me know when it's ready to be merged!

spawnia commented on src/SentryLink.ts:
```diff
@@ -1,225 +1,113 @@
-    this.fillBreadcrumb(breadcrumb, operation);
-
-    // Start observing the operation for results
-    return new Observable<FetchResult>((observer) => {
```
I thought this wrapping was totally unnecessary and ripped it out. Turns out it is quite important to guarantee the callbacks run before the error actually hits Sentry. I now commented that section heavily 😄

spawnia commented:
I am going to be battle-testing this for a few more days with the current feature set. How should I best document the changes? How about an `UPGRADE.md` for the breakage?

DiederikvandenB commented:
Upgrade.md sounds good.

spawnia commented:
Did some more testing with it, everything is working fine. I rewrote the commit messages and added both changelog and upgrade guide. Everything should be ready to merge :)

DiederikvandenB commented:
Cheers. Merged & released.

spawnia commented:
Thank you. Next up, support for spans :)
